File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/features/module/modules/render/BlockOverlay.java
package net.ccbluex.liquidbounce.features.module.modules.render;

import net.ccbluex.liquidbounce.event.EventTarget;
import net.ccbluex.liquidbounce.event.Render2DEvent;
import net.ccbluex.liquidbounce.event.Render3DEvent;
import net.ccbluex.liquidbounce.features.module.Module;
import net.ccbluex.liquidbounce.features.module.ModuleCategory;
import net.ccbluex.liquidbounce.features.module.ModuleInfo;
import net.ccbluex.liquidbounce.ui.font.Fonts;
import net.ccbluex.liquidbounce.utils.block.BlockUtils;
import net.ccbluex.liquidbounce.utils.render.ColorUtils;
import net.ccbluex.liquidbounce.utils.render.RenderUtils;
import net.ccbluex.liquidbounce.value.BoolValue;
import net.ccbluex.liquidbounce.value.IntegerValue;
import net.minecraft.block.Block;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.BlockPos;

import java.awt.Color;

import static net.minecraft.client.renderer.GlStateManager.*;
import static net.minecraft.client.renderer.RenderGlobal.drawSelectionBoundingBox;
import static org.lwjgl.opengl.GL11.*;

/**
 * LiquidBounce Hacked Client
 * A minecraft forge injection client using Mixin
 *
 * @game Minecraft
 * @author CCBlueX
 */
@ModuleInfo(name = "BlockOverlay", description = "Allows you to change the design of the block overlay.", category = ModuleCategory.RENDER)
public class BlockOverlay extends Module {

    private final IntegerValue colorRedValue = new IntegerValue("R", 68, 0, 255);
    private final IntegerValue colorGreenValue = new IntegerValue("G", 117, 0, 255);
    private final IntegerValue colorBlueValue = new IntegerValue("B", 255, 0, 255);
    private final BoolValue colorRainbow = new BoolValue("Rainbow", false);

    public final BoolValue infoValue = new BoolValue("Info", false);

    @EventTarget
    public void onRender3D(Render3DEvent event) {
        if (mc.objectMouseOver != null && mc.objectMouseOver.getBlockPos() != null &&
                mc.theWorld.getBlockState(mc.objectMouseOver.getBlockPos()).getBlock() != null &&
                BlockUtils.canBeClicked(mc.objectMouseOver.getBlockPos())) {
            float partialTicks = event.getPartialTicks();

            enableBlend();
            tryBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA, GL_ONE, GL_ZERO);
            final Color color = colorRainbow.get() ? ColorUtils.rainbow(0.4F) : new Color(colorRedValue.get(), colorGreenValue.get(), colorBlueValue.get(), (int) (0.4F * 255));
            RenderUtils.glColor(color);
            glLineWidth(2F);
            disableTexture2D();

            depthMask(false);
            final BlockPos blockPos = mc.objectMouseOver.getBlockPos();
            final Block block = mc.theWorld.getBlockState(blockPos).getBlock();

            block.setBlockBoundsBasedOnState(mc.theWorld, blockPos);
            final double x = mc.thePlayer.lastTickPosX + (mc.thePlayer.posX - mc.thePlayer.lastTickPosX) * (double) partialTicks;
            final double y = mc.thePlayer.lastTickPosY + (mc.thePlayer.posY - mc.thePlayer.lastTickPosY) * (double) partialTicks;
            final double z = mc.thePlayer.lastTickPosZ + (mc.thePlayer.posZ - mc.thePlayer.lastTickPosZ) * (double) partialTicks;

            final AxisAlignedBB axisAlignedBB = block.getSelectedBoundingBox(mc.theWorld, blockPos)
                    .expand(0.0020000000949949026D, 0.0020000000949949026D, 0.0020000000949949026D)
                    .offset(-x, -y, -z);

            drawSelectionBoundingBox(axisAlignedBB);
            RenderUtils.drawFilledBox(axisAlignedBB);

            depthMask(true);
            enableTexture2D();

            disableBlend();
            GlStateManager.resetColor();
        }
    }

    @EventTarget
    public void onRender2D(final Render2DEvent event) {
        if(mc.objectMouseOver != null && mc.objectMouseOver.getBlockPos() != null && mc.theWorld.getBlockState(mc.objectMouseOver.getBlockPos()).getBlock() != null && BlockUtils.canBeClicked(mc.objectMouseOver.getBlockPos()) && infoValue.get()) {
            final BlockPos blockPos = mc.objectMouseOver.getBlockPos();
            final Block block = mc.theWorld.getBlockState(blockPos).getBlock();

            if(mc.theWorld.getWorldBorder().contains(blockPos)) {
                final String info = block.getLocalizedName() + " §7ID: " + Block.getIdFromBlock(block);
                final ScaledResolution scaledResolution = new ScaledResolution(mc);
                RenderUtils.drawBorderedRect((scaledResolution.getScaledWidth() / 2) - 2, (scaledResolution.getScaledHeight() / 2) + 5, (scaledResolution.getScaledWidth() / 2) + Fonts.font40.getStringWidth(info) + 2, (scaledResolution.getScaledHeight() / 2) + 16, 3, Color.BLACK.getRGB(), Color.BLACK.getRGB());
                GlStateManager.resetColor();
                Fonts.font40.drawString(info, scaledResolution.getScaledWidth() / 2, scaledResolution.getScaledHeight() / 2 + 7, Color.WHITE.getRGB());
            }
        }
    }
}

File: shared/main/java/net/ccbluex/liquidbounce/features/module/modules/render/BlockOverlay.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.features.module.modules.render

import net.ccbluex.liquidbounce.api.minecraft.util.WBlockPos
import net.ccbluex.liquidbounce.api.minecraft.world.IWorld
import net.ccbluex.liquidbounce.event.EventTarget
import net.ccbluex.liquidbounce.event.Render2DEvent
import net.ccbluex.liquidbounce.event.Render3DEvent
import net.ccbluex.liquidbounce.features.module.Module
import net.ccbluex.liquidbounce.features.module.ModuleCategory
import net.ccbluex.liquidbounce.features.module.ModuleInfo
import net.ccbluex.liquidbounce.injection.backend.Backend
import net.ccbluex.liquidbounce.ui.font.Fonts
import net.ccbluex.liquidbounce.utils.extensions.canBeClicked
import net.ccbluex.liquidbounce.utils.extensions.getBlock
import net.ccbluex.liquidbounce.utils.render.ColorUtils.rainbowRGB
import net.ccbluex.liquidbounce.utils.render.RenderUtils
import net.ccbluex.liquidbounce.value.*
import org.lwjgl.opengl.GL11

@ModuleInfo(name = "BlockOverlay", description = "Allows you to change the design of the block overlay.", category = ModuleCategory.RENDER)
class BlockOverlay : Module()
{
    private val colorValue = RGBAColorValue("Color", 255, 255, 255, 30, listOf("R", "G", "B", "Alpha"))

    private val colorRainbowGroup = ValueGroup("Rainbow")
    private val colorRainbowEnabledValue = BoolValue("Enabled", true, "Rainbow")
    private val colorRainbowSpeedValue = IntegerValue("Speed", 10, 1, 10, "Rainbow-Speed")
    private val colorRainbowSaturationValue = FloatValue("Saturation", 1.0f, 0.0f, 1.0f, "HSB-Saturation")
    private val colorRainbowBrightnessValue = FloatValue("Brightness", 1.0f, 0.0f, 1.0f, "HSB-Brightness")

    private val drawHydraESPValue = BoolValue("Hydra", false)

    private val infoGroup = ValueGroup("Info")
    val infoEnabledValue = BoolValue("Enabled", false, "Info")
    private val infoFontValue = FontValue("Font", Fonts.font40)

    private val interpolateValue = BoolValue("Interpolate", true)

    init
    {
        colorRainbowGroup.addAll(colorRainbowEnabledValue, colorRainbowSpeedValue, colorRainbowSaturationValue, colorRainbowBrightnessValue)
        infoGroup.addAll(infoEnabledValue, infoFontValue)
    }

    fun getCurrentBlock(theWorld: IWorld): WBlockPos? = mc.objectMouseOver?.blockPos?.let { if (theWorld.canBeClicked(it) && it in theWorld.worldBorder) it else null }

    @EventTarget
    fun onRender3D(event: Render3DEvent)
    {
        val theWorld = mc.theWorld ?: return
        val thePlayer = mc.thePlayer ?: return

        val blockPos = getCurrentBlock(theWorld) ?: return

        val lastTickPosX = thePlayer.lastTickPosX
        val lastTickPosY = thePlayer.lastTickPosY
        val lastTickPosZ = thePlayer.lastTickPosZ

        val block = theWorld.getBlockState(blockPos).block

        val partialTicks = if (interpolateValue.get()) event.partialTicks else 1f

        val rainbowSpeed = colorRainbowSpeedValue.get()
        val color = if (colorRainbowEnabledValue.get()) rainbowRGB(alpha = colorValue.getAlpha(), speed = rainbowSpeed, saturation = colorRainbowSaturationValue.get(), brightness = colorRainbowBrightnessValue.get()) else colorValue.get()

        val glStateManager = classProvider.glStateManager

        glStateManager.enableBlend()
        glStateManager.tryBlendFuncSeparate(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA, GL11.GL_ONE, GL11.GL_ZERO)
        RenderUtils.glColor(color)
        GL11.glLineWidth(2F)
        glStateManager.disableTexture2D()
        GL11.glDepthMask(false)

        @Suppress("ConstantConditionIf") if (Backend.MINECRAFT_VERSION_MINOR < 12) block.setBlockBoundsBasedOnState(theWorld, blockPos)

        val x = lastTickPosX + (thePlayer.posX - lastTickPosX) * partialTicks
        val y = lastTickPosY + (thePlayer.posY - lastTickPosY) * partialTicks
        val z = lastTickPosZ + (thePlayer.posZ - lastTickPosZ) * partialTicks

        val boxExpandSize = 0.002
        val axisAlignedBB = block.getSelectedBoundingBox(theWorld, theWorld.getBlockState(blockPos), blockPos).expand(boxExpandSize, boxExpandSize, boxExpandSize).offset(-x, -y, -z)

        RenderUtils.drawSelectionBoundingBox(axisAlignedBB, drawHydraESPValue.get())
        RenderUtils.drawFilledBox(axisAlignedBB)
        GL11.glDepthMask(true)

        glStateManager.enableTexture2D()
        glStateManager.disableBlend()
        glStateManager.resetColor()
    }

    @EventTarget
    fun onRender2D(@Suppress("UNUSED_PARAMETER") event: Render2DEvent)
    {
        if (infoEnabledValue.get())
        {
            val theWorld = mc.theWorld ?: return

            val blockPos = getCurrentBlock(theWorld) ?: return
            val block = theWorld.getBlock(blockPos)

            val info = "${block.localizedName} \u00A77ID: ${functions.getIdFromBlock(block)}"

            val provider = classProvider

            val scaledResolution = provider.createScaledResolution(mc)

            val font = infoFontValue.get()
            val middleScreenX = scaledResolution.scaledWidth shr 1
            val middleScreenY = scaledResolution.scaledHeight shr 1

            RenderUtils.drawBorderedRect(middleScreenX - 2F, middleScreenY + 5F, (middleScreenX + font.getStringWidth(info)) + 2F, middleScreenY + 16F, 3F, -16777216, -16777216)

            provider.glStateManager.resetColor()
            font.drawString(info, middleScreenX.toFloat(), middleScreenY + 7f, 0xffffff, false)
        }
    }
}
